Improve Plugwise tests

CoMPaTech opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
Follow-up from https://github.com/home-assistant/core/pull/131888 review comments.

@joostlek reviewing our `config_flow` and attached testing I don't see any misses for `unique-config-entry` & `config-flow-test-coverage` unless I overlook something completely.

- [x] Config flow applies 4 aborts on zeroconf, 1 on user entry and an additional for trying to configure an Anna when part of zonecontrol (i.e. Adam) setup
- [x] Both `test_form` and `test_zeroflow_form` are present
- [x] As an example `test_zeroconf_abort_anna_with_existing_config_entries` is there to check duplicity


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #131888
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

CoMPaTech committed:
Update after verification

joostlek commented:
https://github.com/home-assistant/core/pull/131888#discussion_r1865520653
https://github.com/home-assistant/core/pull/131888#discussion_r1865522889
> As an example test_zeroconf_abort_anna_with_existing_config_entries is there to check duplicity

But that's only the zeroconf flow, we should also explicitly test this in the user flow

Also, `test_zeroflow_form` -> It's zeroconf, not zeroflow

CoMPaTech committed:
Add tests accordingly

CoMPaTech commented:
> [#131888 (comment)](https://github.com/home-assistant/core/pull/131888#discussion_r1865520653) [#131888 (comment)](https://github.com/home-assistant/core/pull/131888#discussion_r1865522889)
Added accordingly


> > As an example test_zeroconf_abort_anna_with_existing_config_entries is there to check duplicity
> But that's only the zeroconf flow, we should also explicitly test this in the user flow
 Added to both flows, not only the abort on the Anna/Adam construct

> Also, `test_zeroflow_form` -> It's zeroconf, not zeroflow
Apologies for the typo!

joostlek commented on tests/components/plugwise/test_config_flow.py:
```diff
@@ -128,6 +129,26 @@ async def test_form(
     assert len(mock_setup_entry.mock_calls) == 1
     assert len(mock_smile_config_flow.connect.mock_calls) == 1
 
+    entries = hass.config_entries.async_entries(config_flow.DOMAIN)
+    assert entries[0].unique_id == "smile12345"
```
```suggestion
    assert result2["result"].unique_id == "smile12345"
```

joostlek replied:
But we don't ask `async_entries`, the created config entry is returned in the config_entries.async_configure.

CoMPaTech replied:
Sorry, misread that, adjusted accordingly

joostlek commented on tests/components/plugwise/test_config_flow.py:
```diff
@@ -128,6 +129,26 @@ async def test_form(
+    entries = hass.config_entries.async_entries(config_flow.DOMAIN)
+    assert entries[0].unique_id == "smile12345"
+
+    # Reinitialize flow
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={CONF_SOURCE: SOURCE_USER}
+    )
+
+    result3 = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        user_input={
+            CONF_HOST: TEST_HOST,
+            CONF_PASSWORD: TEST_PASSWORD,
+        },
+    )
+    await hass.async_block_till_done()
+
+    assert result3.get("type") is FlowResultType.ABORT
+    assert result3.get("reason") == "already_configured"
```
Let's test this in a separate test and by adding the MockConfigEntry and then seeing it abort

CoMPaTech replied:
Added as separate test now

joostlek commented on tests/components/plugwise/test_config_flow.py:
```diff
@@ -172,6 +193,25 @@ async def test_zeroconf_flow(
     assert len(mock_smile_config_flow.connect.mock_calls) == 1
 
+    entries = hass.config_entries.async_entries(config_flow.DOMAIN)
+    assert entries[0].unique_id == "smile12345"
```
Also use that other method

CoMPaTech replied:
👍

joostlek commented on tests/components/plugwise/test_config_flow.py:
```diff
@@ -172,6 +193,25 @@ async def test_zeroconf_flow(
+    entries = hass.config_entries.async_entries(config_flow.DOMAIN)
+    assert entries[0].unique_id == "smile12345"
+
+    # Reinitialize flow
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={CONF_SOURCE: SOURCE_ZEROCONF},
+        data=TEST_DISCOVERY,
+    )
+
+    result3 = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        user_input={CONF_PASSWORD: TEST_PASSWORD},
+    )
+    await hass.async_block_till_done()
+
+    assert result3.get("type") is FlowResultType.ABORT
+    assert result3.get("reason") == "already_configured"
```
This is already tested by `test_zeroconf_abort_anna_with_existing_config_entries` no?

CoMPaTech replied:
On hindsight that test also revolves around trying to configure an Anna when a Adam is already present.

CoMPaTech replied:
@bouwew please confirm, but @joostlek the `test_zeroconf_abort_anna_with_existing_config_entries` tests for configuring Anna while Adam already is present while `test_zeroconf_abort_anna_with_adam` tests aborting configuring an Anna from discovery when an Adam is also discovered

joostlek replied:
If there are more combinations that we should test, let's do that then, as long as it is tested :)

CoMPaTech replied:
There are 3 now at the bottom - the original two for the adam/anna combination as mentioned + the additional one for validating through the mocked configentry as part of the quoted/reviewed code above

CoMPaTech replied:
Added accordingly

bouwew replied:
@CoMPaTech confirm: `test_zeroconf_abort_anna_with_existing_config_entries` tests that discovering the Anna is blocked when the Adam is already present. The other test does what you write :)

CoMPaTech committed:
Rework from feedback

CoMPaTech committed:
Add user abort test

CoMPaTech commented:
Should be all good now (tnx @bouwew for confirming and @joostlek for your patience)

CoMPaTech commented:
Tnx @joostlek 👍
